### PR TITLE
[NPU] Add Ascend installation manifest and helper

### PR DESCRIPTION
**Pull request by wangyao-i**
## Motivation

The default `pyproject.toml` includes CUDA-specific dependencies that can conflict with an existing Ascend `torch_npu` environment. This PR adds a separate, maintainable installation path for Ascend NPU without changing the externally managed accelerator software stack.

Part of #1597.

## Modifications

- Add `pyproject_npu.toml` with an Ascend-compatible dependency set and optional extras while leaving `torch`, `torch_npu`, `triton-ascend`, `sgl-kernel-npu`, and SGLang under user/upstream version management.
- Align the documented NPU installation baseline with the SGLang `0.5.18` release line.
- Keep the supported SGLang release in one configuration value so future upgrades do not require maintaining a duplicated upper-bound version.
- Add executable `scripts/npu/install_npu.sh`, organized into focused argument parsing, precheck, install, cleanup, verification, and `main` functions.
- Add prechecks for:
  - installed SGLang belonging to the supported `0.5.18` release line, including development, pre-release, final, post-release, and local builds such as `0.5.18.dev7+g<git-sha>`
  - `torch`/`torch_npu` major-minor compatibility
  - `triton-ascend` and `sgl-kernel-npu` imports
  - NPU availability/count and a minimal MatMul
  - `--skip-device-check` for build hosts without exposed devices
- Preserve the original CUDA manifest with locking, stale-backup detection, and restore-on-exit behavior.
- Add `docs/get_started/installation_npu.md`, distinguishing required/manual prerequisites, linking to authoritative compatibility documentation, and providing one recommended installation path.
- Expose the NPU guide through the Get Started toctree and general installation page.
- Add install-script tests for supported SGLang release-line variants, mismatch and missing cases, interrupted recovery, locking, dry-run behavior, extras, build-isolation defaults, and executable paths containing spaces.
- Use the unique test filename `test_npu_install_script.py` to avoid pytest import collisions with the existing XPU installer tests.

## Validation

Completed locally:

- Relevant pre-commit hooks pass, including the executable-shebang check.
- `bash -n scripts/npu/install_npu.sh`
- `pyproject_npu.toml` parses successfully.
- `python -m pytest tests/unit_test/npu/test_npu_install_script.py -q -k "not second_run_refuses_while_lock_is_held"`: 21 passed, 1 deselected.
- Joint collection of the NPU and XPU installer suites: 24 tests collected without module-name conflicts.
- `python .github/update_ci_permission.py --sort-only`: already sorted, no changes.
- `git diff --check`

The local host is Windows and does not provide a real Linux `flock` or Ascend NPU environment. Run the final Linux/Ascend validation with:

```bash
source /usr/local/Ascend/ascend-toolkit/set_env.sh
pytest -q tests/unit_test/npu/test_npu_install_script.py
bash scripts/npu/install_npu.sh --check
bash scripts/npu/install_npu.sh
sgl-omni --help
```
## Accuracy Test

N/A — packaging, installation, and documentation only.

## Benchmark & Profiling

N/A — no inference-path changes.

## Checklist

- [x] Format code according to pre-commit conventions.
- [x] Add unit tests.
- [x] Update documentation.
- [ ] Run the full Linux/Ascend validation commands above and attach the output.

### CODE REVIEW

File: docs/get_started/installation_npu.md [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+`pip install -e .` resolves the CUDA [`pyproject.toml`](../../pyproject.toml), whose torch
+family and CUDA-only wheels would replace the `torch+npu` stack.
+[`pyproject_npu.toml`](../../pyproject_npu.toml) encodes the NPU replacements.
+
+Core deps cover the supported models (Qwen3-ASR / TTS / Omni) plus the API server;
+`[eval]` adds SeedTTS/WER tooling, `[all]` aliases it, and `[fun-cosyvoice3]`
+adds the Fun-CosyVoice3 dependencies. Other model families
+(S2-Pro, Ming-Omni, Voxtral-TTS) are CUDA-only and are not offered here.
+
+> **`--no-build-isolation` is required** — without it pip emits a legacy in-tree
```
Comment (Makcum888e) [inline docs/get_started/installation_npu.md]: `--no-build-isolation` do you need this?

File: docs/get_started/installation_npu.md [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+```bash
+sgl-omni serve --model-path /path/to/Qwen3-TTS-12Hz-1.7B-Base --host 0.0.0.0 --port 8000
+# Base checkpoint clones a reference voice — pass ref_audio (+ ref_text):
+curl -s -X POST http://localhost:8000/v1/audio/speech \
+  -H "Content-Type: application/json" \
+  -d '{"model":"/path/to/Qwen3-TTS-12Hz-1.7B-Base","input":"Hello from Ascend NPU.",
+       "voice":"default","ref_audio":"/path/to/ref.wav","ref_text":"reference transcript",
+       "response_format":"wav"}' -o out.wav
+```
```
Comment (Makcum888e) [inline docs/get_started/installation_npu.md]: Have you checked that these models work?

File: docs/get_started/installation_npu.md (hunk 1 above)
Comment (wangyao-i) [reply docs/get_started/installation_npu.md]: Thanks for pointing this out. It is not needed for installing sglang-omni. I verified the editable install in a clean virtual environment with build isolation enabled: pip successfully built a PEP 660 editable wheel, and the package was importable outside the repository. I removed the flag and the related setuptools preflight in 08f49ba7.

File: docs/get_started/installation_npu.md (hunk 2 above)
Comment (wangyao-i) [reply docs/get_started/installation_npu.md]: Only the Qwen3-Omni text-only path has been validated on Ascend NPU. Qwen3-ASR and Qwen3-TTS have not been validated on NPU yet, so I removed those serving examples and narrowed the documented support scope in 08f49ba7.

File: scripts/npu/install_npu.sh [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+errors = []
+torch = None
+torch_npu = None
+
+try:
+    import torch
+    print(f"  torch:       {torch.__version__}")
+except ImportError:
+    errors.append("torch is not installed. Install CPU torch first:\n"
+                  "  pip install torch==2.10.0 torchvision==0.25.0 --index-url https://download.pytorch.org/whl/cpu")
```
Comment (ping1jing2) [inline scripts/npu/install_npu.sh]: torch2.10.0 is just an example, please post [Ascend for PyTorch](https://www.hiascend.com/developer/software/ai-frameworks/pytorch/download?versionId=177&ids=89dda9ba9de741349efa03687a487678%2C204%2C200%2C1%2C6%2C177%2C) link here. including the below similar cases

File: scripts/npu/install_npu.sh [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+EDITABLE="-e"
+CHECK_ONLY=0
+SKIP_DEVICE_CHECK=0
+EXTRAS=""
+
+while [[ $# -gt 0 ]]; do
+  case "$1" in
+    --no-editable) EDITABLE=""; shift ;;
+    --check)       CHECK_ONLY=1; shift ;;
+    --skip-device-check) SKIP_DEVICE_CHECK=1; shift ;;
```
Comment (ping1jing2) [inline scripts/npu/install_npu.sh]: the whole shell script is hard to read and understand, please re-organize this script. for example: `usage()`, `precheck()`, `install()`, `main()` and so on.

File: docs/get_started/installation_npu.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+## Why a separate pyproject
+
+`pip install -e .` resolves the CUDA [`pyproject.toml`](../../pyproject.toml), whose torch
+family and CUDA-only wheels would replace the `torch+npu` stack.
+[`pyproject_npu.toml`](../../pyproject_npu.toml) encodes the NPU replacements.
+
+Core dependencies cover Qwen3-Omni plus the API server;
+`[eval]` adds SeedTTS/WER tooling, `[all]` aliases it, and `[fun-cosyvoice3]`
+adds the Fun-CosyVoice3 dependencies. Other model families
+(S2-Pro, Ming-Omni, Voxtral-TTS) are CUDA-only and are not offered here.
```
Comment (ping1jing2) [inline docs/get_started/installation_npu.md]: please delete this part, just need to tell users how to install sglang-omni on Ascend server directly

File: docs/get_started/installation_npu.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+| Component | Version | Install |
+|-----------|---------|---------|
+| CANN toolkit | 9.0.0 | [Installation guide](https://www.hiascend.com/document/detail/zh/CANNCommunityEdition/900/softwareinst/instg/instg_0008.html) |
+| HDK (driver/firmware) | 25.5.2 | [Download](https://www.hiascend.com/hardware/firmware-drivers/community) |
+| torch (CPU) | 2.10.0 | `pip install torch==2.10.0 torchvision==0.25.0 torchaudio==2.10.0 --index-url https://download.pytorch.org/whl/cpu` |
+| torch_npu | 2.10.0 | [gitcode.com/Ascend/pytorch/releases](https://gitcode.com/Ascend/pytorch/releases) (arch-specific wheel) |
+| triton-ascend | 3.2.1.dev20260530 | `pip install triton-ascend==3.2.1.dev20260530 --extra-index-url=https://mirrors.huaweicloud.com/ascend/repos/pypi/nightly --trusted-host mirrors.huaweicloud.com` |
+| memfabric-hybrid | 1.0.8 | `pip install memfabric-hybrid==1.0.8` (PD disaggregation only) |
+| sgl-kernel-npu | latest | [sgl-project/sgl-kernel-npu](https://github.com/sgl-project/sgl-kernel-npu) releases |
```
Comment (ping1jing2) [inline docs/get_started/installation_npu.md]: 1. please add one more column named `required` or `manul install`, especially fill it with `yes` or `no`. if `install_npu.sh` can handle it automatically, please describe it explicitly.
2. using `installation` replace `install`, and write `[official documentation](hiasend link)` uniformly, delete all the pip command.

File: docs/get_started/installation_npu.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+`torch_npu` wheels are arch-specific and hosted on Huawei's gitcode mirror, not on PyPI.
+Download the matching wheel for your architecture:
+
+```bash
+# aarch64 (Atlas 800I A2 / A3)
+pip install https://gitcode.com/Ascend/pytorch/releases/download/v26.0.0-pytorch2.10.0/torch_npu-2.10.0-cp311-cp311-manylinux_2_28_aarch64.whl
+
+# x86_64
+pip install https://gitcode.com/Ascend/pytorch/releases/download/v26.0.0-pytorch2.10.0/torch_npu-2.10.0-cp311-cp311-manylinux_2_28_x86_64.whl
+```
```
Comment (ping1jing2) [inline docs/get_started/installation_npu.md]: we don't need it as we have `install_npu.sh` and official link

File: docs/get_started/installation_npu.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+```bash
+git clone https://github.com/sgl-project/sglang-omni.git
+cd sglang-omni
+source /usr/local/Ascend/ascend-toolkit/set_env.sh
+
+# dry-run first — shows the commands, installs nothing
+PYTHON=$(which python) scripts/npu/install_npu.sh --check
+
+# editable install
+PYTHON=$(which python) scripts/npu/install_npu.sh
```
Comment (ping1jing2) [inline docs/get_started/installation_npu.md]:
```suggestion
python3 scripts/npu/install_npu.sh --check

# editable install
python3 scripts/npu/install_npu.sh
```

File: docs/get_started/installation_npu.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
+# 🚀 Installation — Huawei Ascend NPU
```
Comment (ping1jing2) [inline docs/get_started/installation_npu.md]: keep the documentation simple and ease to use, delete the information if already exist in `install_npu.sh`

**Review by Makcum888e — APPROVED** [review]
LGTM

File: docs/get_started/installation_npu.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
+# Installation — Huawei Ascend NPU
```
Comment (ping1jing2) [inline docs/get_started/installation_npu.md]: delete `Huawei` in the whole SGLang related codebase

File: docs/get_started/installation_npu.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+Install the Ascend software stack and NPU build of SGLang before installing
+`sglang-omni`. The helper script installs only `sglang-omni`; it does not install
```
Comment (ping1jing2) [inline docs/get_started/installation_npu.md]: i think `install_npu.sh` is _The helper script_. if so, please write it like `The helper script [install_npu.sh](link)`

File: docs/get_started/installation_npu.md [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+# Install sglang-omni in editable mode.
+bash scripts/npu/install_npu.sh
+```
+
+The precheck verifies the required Python packages, matching `torch` and
+`torch_npu` major-minor versions, NPU availability, and a small NPU matrix
+multiplication. Run `bash scripts/npu/install_npu.sh --help` for optional extras,
+non-editable installation, and environments where devices are intentionally not
+exposed during the build.
```
Comment (ping1jing2) [inline docs/get_started/installation_npu.md]: we need `verify` part at the end

**Comment by ping1jing2** [comment]
please fix the lint issue
https://github.com/sgl-project/sglang-omni/actions/runs/33349427554/job/99402636299?pr=1577
```
scripts/npu/install_npu.sh: has a shebang but is not marked executable!
  If it is supposed to be executable, try: `chmod +x scripts/npu/install_npu.sh`
  If on Windows, you may also need to: `git add --chmod=+x scripts/npu/install_npu.sh`
  If it not supposed to be executable, double-check its shebang is wanted.
```

File: scripts/npu/install_npu.sh [hunk 13]
```diff
@@ -0,0 +1,347 @@
...
+# Install sglang-omni against a pre-installed Huawei Ascend NPU stack.
+
+set -euo pipefail
+
+REPO_ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")/../.." && pwd)"
+PYPROJECT="${REPO_ROOT}/pyproject.toml"
+PYPROJECT_NPU="${REPO_ROOT}/pyproject_npu.toml"
+BACKUP="${REPO_ROOT}/.pyproject.cuda.bak"
+LOCK="${REPO_ROOT}/.pyproject.npu.lock"
+SGLANG_VERIFIED_VERSION="v0.5.16"
```
Comment (Ratish1) [inline scripts/npu/install_npu.sh]: current main requires sglang 0.5.18, but this helper and guide still select v0.5.16 and allow a missing sglang install to pass verification. update every reference to v0.5.18 and fail the precheck when sglang is missing or reports a different version.

**Comment by ping1jing2** [comment]
please double check and update your PR description

File: scripts/npu/install_npu.sh [hunk 14]
```diff
@@ -0,0 +1,347 @@
...
+TRITON_ASCEND_URL = (
+    "https://gitcode.com/Ascend/triton-ascend/blob/main/docs/en/quick_start.md"
+)
+SGL_KERNEL_NPU_URL = "https://github.com/sgl-project/sgl-kernel-npu/releases"
+
+errors = []
+torch = None
+torch_npu = None
+
+try:
```
Comment (Ratish1) [inline scripts/npu/install_npu.sh]: qwen3-omni resolves `sglang_omni.preprocessing.video` during preprocessor import, which imports `torchvision` even in text-only mode. the sglang v0.5.18 npu setup installs torchvision with the torch stack; add it to the prerequisites and fail the precheck when it is not importable.

File: pyproject_npu.toml [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+override-dependencies = [
+    "protobuf>=6.31.1,<7.0.0",
+]
+
+[tool.pytest.ini_options]
+pythonpath = ["."]
+markers = [
+    "benchmark: performance benchmark tests (may require GPU and long runtime)",
+    "tts_stage(name): select an in-file TTS benchmark CI stage",
+    "gpu: requires an accelerator device (auto-skipped without one)",
```
Comment (Ratish1) [inline pyproject_npu.toml]: nit: change this to `accelerator: requires accelerator hardware`; the repository uses `pytest.mark.accelerator`, and the other manifests register that marker.

File: scripts/npu/install_npu.sh [hunk 16]
```diff
@@ -0,0 +1,347 @@
+#!/usr/bin/env bash
```
Comment (Ratish1) [inline scripts/npu/install_npu.sh]: mark this script executable with `git add --chmod=+x scripts/npu/install_npu.sh`; the current lint check fails because the file has a shebang but mode 100644.

File: scripts/npu/install_npu.sh [hunk 17]
```diff
@@ -0,0 +1,347 @@
...
+# Install sglang-omni against a pre-installed Ascend NPU stack.
+
+set -euo pipefail
+
+REPO_ROOT="$(cd "$(dirname "${BASH_SOURCE[0]}")/../.." && pwd)"
+PYPROJECT="${REPO_ROOT}/pyproject.toml"
+PYPROJECT_NPU="${REPO_ROOT}/pyproject_npu.toml"
+BACKUP="${REPO_ROOT}/.pyproject.cuda.bak"
+LOCK="${REPO_ROOT}/.pyproject.npu.lock"
+SGLANG_VERIFIED_VERSION="v0.5.16"
```
Comment (Ratish1) [inline scripts/npu/install_npu.sh]: current main requires sglang 0.5.18, but we mention v0.5.16 here. can we update every reference to v0.5.18.

File: docs/get_started/installation_npu.md [hunk 18]
```diff
@@ -0,0 +1,41 @@
+# Installation — Ascend NPU
```
Comment (Ratish1) [inline docs/get_started/installation_npu.md:1]: add this page to the get started toctree in `docs/index.rst` and link it from `docs/get_started/installation.md`; otherwise the page builds but is not reachable from the published installation docs.

Comment (wangyao-i) [reply docs/get_started/installation_npu.md:1]: Added installation_npu.md to the Get Started toctree in docs/index.rst and linked it from docs/get_started/installation.md in cb02697e. The NPU installation guide is now reachable through the published documentation navigation.

File: scripts/npu/install_npu.sh (hunk 17 above)
Comment (wangyao-i) [reply scripts/npu/install_npu.sh]: Updated all NPU installation references from SGLang v0.5.16 to v0.5.18, including install_npu.sh, pyproject_npu.toml, and installation_npu.md, in cb02697e. The installation has also been successfully validated on an Ascend server.

**Review by Ratish1 — APPROVED** [review]
LGTM

**Comment by ping1jing2** [comment]
LGTM
